src/xsens_streaming/udpserver.cpp

```diff
@@ -63,11 +62,78 @@ void UdpServer::readMessages()
-      if(datagram->messageType() == StreamingProtocol::SPPoseQuaternion)
+      switch(datagram->messageType())
       {
-        auto & quaternionDatagram = dynamic_cast<QuaternionDatagram &>(*datagram);
-        std::lock_guard<std::mutex> lock(quaternionMutex_);
-        quaternions_ = quaternionDatagram.data();
+        case StreamingProtocol::SPPoseQuaternion:
+        {
+          auto & quaternionDatagram = dynamic_cast<QuaternionDatagram &>(*datagram);
+          std::lock_guard<std::mutex> lock(quaternionMutex_);
+          quaternions_ = quaternionDatagram.data();
+          break;
+        }
+        case StreamingProtocol::SPJointAngles:
+        {
+          auto & jointAnglesDatagram = dynamic_cast<JointAnglesDatagram &>(*datagram);
+          std::lock_guard<std::mutex> lock(jointAnglesMutex_);
+          jointAngles_ = jointAnglesDatagram.data();
+          break;
+        }
+        case StreamingProtocol::SPPoseEuler:
+        {
+          auto & eulerDatagram = dynamic_cast<EulerDatagram &>(*datagram);
+          std::lock_guard<std::mutex> lock(eulerMutex_);
+          euler_ = eulerDatagram.data();
+          break;
+        }
+        case StreamingProtocol::SPPosePositions:
+        {
+          auto & positionDatagram = dynamic_cast<PositionDatagram &>(*datagram);
+          std::lock_guard<std::mutex> lock(virtualMarkerPositionMutex_);
+          virtualMarkerPositions_ = positionDatagram.data();
+          break;
+        }
+        case StreamingProtocol::SPLinearSegmentKinematics:
+        {
+          auto & linearSegmentKinematicsDatagram = dynamic_cast<LinearSegmentKinematicsDatagram &>(*datagram);
+          std::lock_guard<std::mutex> lock(linearSegmentKinematicsMutex_);
+          linearSegmentKinematics_ = linearSegmentKinematicsDatagram.data();
+          break;
+        }
+        case StreamingProtocol::SPAngularSegmentKinematics:
+        {
+          auto & angularSegmentKinematicsDatagram = dynamic_cast<AngularSegmentKinematicsDatagram &>(*datagram);
+          std::lock_guard<std::mutex> lock(angularSegmentKinematicsMutex_);
+          angularSegmentKinematics_ = angularSegmentKinematicsDatagram.data();
+          break;
+        }
+        case StreamingProtocol::SPTrackerKinematics:
+        {
+          auto & trackerKinematicsDatagram = dynamic_cast<TrackerKinematicsDatagram &>(*datagram);
+          std::lock_guard<std::mutex> lock(trackerDataMutex_);
+          trackerData_ = trackerKinematicsDatagram.data();
+          break;
+        }
+        case StreamingProtocol::SPMetaScaling:
+        {
+          auto & scaleDatagram = dynamic_cast<ScaleDatagram &>(*datagram);
+          std::lock_guard<std::mutex> lock1(pointDefinitionMutex_);
+          pointDefinition_ = scaleDatagram.pointDefinition();
+          std::lock_guard<std::mutex> lock2(nullPoseDefinitionMutex_);
+          nullPoseDefinition_ = scaleDatagram.nullPoseDefinition();
```

Comment:
I don't think we should have two locks here: both `pointDefinition` and `nullPoseDefinition` are received in the same datagram. I haven't looked up their meaning, but I think they are related, and thus it would make more sense to retrieve them together.

Reply:
You're right, I changed to a single mutex.